### ISSUE TITLE
Jerk exceeding boundary

### ISSUE THREAD

**Issue by zeyang23** [translearn/limits#2]
Dear @translearn
I got a jerk trajectory exceeding the boundary when using this lib on a single axis.
The parameters are
```
time_step = 0.005
trajectory_duration = 5
pos_limits = [[0, 200], ]
vel_limits = [[-400., 400.], ]
acc_limits = [[-4000., 4000.], ]
jerk_limits = [[-100000., 100000.], ]
```
I set the _action_=1.0, i.e. always utiilze the maximal valid acceleration.
The output is:
![image](https://user-images.githubusercontent.com/50078363/159894983-8ad80cae-1d8b-475f-b8b5-4c1e342ef399.png)
From the pic we can see that at some points the jerk is extremely large. Besides, the oscillations are annoying.
Looking forward for your reply!

**Comment by translearn** [comment]
Dear @zeyang23,
thank you very much for reporting this issue!
I was able to reproduce the behavior by running the following command:

```bash
python klimits/test_trajectory_generation.py --time_step=0.1 --pos_limits='[[0, 200]]' --vel_limits='[[-400, 400]]' --acc_limits='[[-4000, 4000]]' --jerk_limit_factor=0.0625 --time_step=0.005 --trajectory_duration=1.5 --constant_action=1.0
```

![action_1 0_failure](https://user-images.githubusercontent.com/51738372/159950945-57c93383-8289-4d91-9b1e-fd287ecbfd94.png)

The problem is caused by a numerical issue. When setting the action to 1.0, the maximum valid acceleration is selected at each time step. If the computed acceleration is slightly larger than the actual one, no valid solution will be found in a following time step. In this case, the library temporally ignores the jerk limits to make sure that the other limits are not exceeded.
I will look into this issue soon.

In the meantime, you can circumvent the problem by selecting a constant action that is slightly smaller than 1.0, e.g. with --constant_action=0.999 the jerk limits are no longer violated:

```bash
python klimits/test_trajectory_generation.py --time_step=0.1 --pos_limits='[[0, 200]]' --vel_limits='[[-400, 400]]' --acc_limits='[[-4000, 4000]]' --jerk_limit_factor=0.0625 --time_step=0.005 --trajectory_duration=1.5 --constant_action=0.999
```

![action_0 999](https://user-images.githubusercontent.com/51738372/159953316-1d9d5d49-955d-455a-99c3-2a3203298164.png)

Note that the velocity no longer oscillates around its maximum value.
However, due to the discrete decision steps, the acceleration and the jerk still oscillate once the maximum position is reached.
This oscillation will decay quickly if the action is set to a smaller value, e.g. 0.98:

```bash
python klimits/test_trajectory_generation.py --time_step=0.1 --pos_limits='[[0, 200]]' --vel_limits='[[-400, 400]]' --acc_limits='[[-4000, 4000]]' --jerk_limit_factor=0.0625 --time_step=0.005 --trajectory_duration=1.5 --constant_action=0.98
```
![action_0 98](https://user-images.githubusercontent.com/51738372/159957022-9eee71cc-3a4a-47c5-b4a5-a18b941550a9.png)

If you experience problems with oscillations, you can reduce the range of safe accelerations by ~ 2 % using the following code:

```
safe_acc_range_diff = safe_acc_max - safe_acc_min
safe_acc_max = safe_acc_min + 0.99 * safe_acc_range_diff
safe_acc_min = safe_acc_min + 0.01 * safe_acc_range_diff 
```

Please let me know, if you have any further questions!

**Comment by zeyang23** [comment]
Thanks a lot for your detailed explanation! 👍👍👍

**Comment by zeyang23** [comment]
Dear @translearn,
It seems that the oscillation problems happen when smaller time_steps are set. For example, if set time_step=0.02, I can get a perfect trajectory with no oscillation:
![image](https://user-images.githubusercontent.com/50078363/160086353-5975b132-4e3a-45a3-8e35-97ca6fa821a8.png)
However, to get smoother trajectories I have to lower the time_step. If set time_step==0.0002, it seems that no matter how to reduce the range of safe accelerations, the oscillations will occur (5% of the range of safe accelerations is reduced):
![image](https://user-images.githubusercontent.com/50078363/160087417-93311282-b18d-4f20-9348-b108b8b73d93.png)

**Comment by translearn** [comment]
Dear @zeyang23,
thank you very much for your feedback.
I will also look into this.
Note that typical time steps for the library are in the range of 10ms to 100ms (in our reinforcement learning experiments, these time steps led to the best learning performance). 
Our library considers that actions can only be made at discrete decision steps which means that the acceleration values of two following time steps can always be connected by a straight line. 
However, for very small time steps like 0.0002 s, it might make more sense to use a continuous solver like Reflexxes (or the open source implementation Ruckig). To compute the maximum valid acceleration with these libraries, set the target position to p_max and the target velocity (and acceleration) to zero.

**Comment by translearn** [comment]
@zeyang23
Can you provide the exact settings of our experiment above? With time_step=0.0002 and the limits from your initial post, i get the following results:

![action_time_step_0 0002](https://user-images.githubusercontent.com/51738372/160841581-bf22a794-cac5-492e-9be1-1736da7fd547.png)

**Comment by zeyang23** [comment]
Dear @translearn 
I sincerely appreciate your patient reply. Hope I'm not disturbing you.
You can run the _test.py_ script to reproduce the behavior I mentioned above.
[test.zip](https://github.com/translearn/limits/files/8381765/test.zip)

**Comment by translearn** [comment]
Hey @zeyang23,
running your code I found that the oscillation is decaying, however the decay takes about 8s:
![test_py_10s](https://user-images.githubusercontent.com/51738372/162981540-0d060d2a-c7fd-4393-aff0-a603601a6498.png)

The reason for the oscillation is that the jerk limit is very low compared to the maximum jerk resulting from the acceleration limits and the selected time step. In your case, the maximum jerk is (a_max - a_min) / time_step = 40 000 000, 
while the jerk limit is 100 000 (0.25 % compared to the maximum jerk).

You can avoid this kind of oscillation by selecting a larger time step. Note that you can still interpolate intermediate setpoints at a higher frequency (e.g. for a trajectory controller):

```python
time_since_start = np.linspace(time_step / 100, self._time_step, 100)  # 100 intermediate setpoints
intermediate_acceleration = interpolate_acceleration_batch(current_acceleration, next_acceleration, time_since_start, time_step)
intermediate_velocity = interpolate_velocity_batch(current_acceleration, next_acceleration, current_velocity, time_since_start, time_step)
intermediate_position = interpolate_position_batch(current_acceleration, next_acceleration, current_velocity, current_position, time_since_start, time_step)
```

**Comment by translearn** [comment]
The initial issue (jerk limit violation) was fixed in version 1.1.2.

**Comment by zeyang23** [comment]
Thank you very much!
